### PR TITLE
zkvm: encapsulate bulletproofs dependencies

### LINKED ISSUE #210: zkvm: remove bulletproofs dependency from API
Would be good if BulletproofsGens and other entities from BP were encapsulated inside ZkVM or at least re-exported, so that the users can add ZkVM as a dependency w/o also bringing in bulletproofs. See `token` crate as an example.

1. `BulletproofsGens`: those can be instantiated by ZkVM with a capacity parameter. R1CS can take them at the `Prover::prove` method, so ZkVM can just create a necessary capacity based on number of multiplicative constraints used. This can also be a part of the `Tx` object.
2. `PedersenGens` are already baked into Commitment and Predicate API, and we can do the same for Signature/VerificationKey API, so users do not create those things with raw scalar operations. If they need to do something custom and unsupported by ZkVM APIs, then they can import Ristretto or Bulletproofs explicitly.
3. R1CS instance is already encapsulated inside ZkVM prover/verifier.

### PR DESCRIPTION
**Pull request by vickiniu**
Fixes #210 - uses capacity of 256 for bulletprooofs generator.

### CODE REVIEW

File: zkvm/src/verifier.rs [hunk 1]
```diff
@@ -79,10 +79,11 @@ impl<'a, 'b> Delegate<r1cs::Verifier<'a, 'b>> for Verifier<'a, 'b> {
 impl<'a, 'b> Verifier<'a, 'b> {
     /// Verifies the `Tx` object by executing the VM and returns the `VerifiedTx`.
     /// Returns an error if the program is malformed or any of the proofs are not valid.
-    pub fn verify_tx<'g>(tx: Tx, bp_gens: &'g BulletproofGens) -> Result<VerifiedTx, VMError> {
+    pub fn verify_tx(tx: Tx) -> Result<VerifiedTx, VMError> {
+        let bp_gens = BulletproofGens::new(128, 1);
```
Comment (oleganza) [inline zkvm/src/verifier.rs]: We cannot hardcode this for two reasons:

1. Bigger txs will have larger circuits and more than 128 multipliers, failing the verification if 128 is hardcoded.
2. Computing generators is expensive. Validator would want to precompute them and use over its whole lifetime to validate arbitrary amount of transactions, so we need to pass it in.

File: zkvm/src/verifier.rs [hunk 2]
```diff
@@ -79,10 +79,11 @@ impl<'a, 'b> Delegate<r1cs::Verifier<'a, 'b>> for Verifier<'a, 'b> {
 impl<'a, 'b> Verifier<'a, 'b> {
     /// Verifies the `Tx` object by executing the VM and returns the `VerifiedTx`.
     /// Returns an error if the program is malformed or any of the proofs are not valid.
-    pub fn verify_tx<'g>(tx: Tx, bp_gens: &'g BulletproofGens) -> Result<VerifiedTx, VMError> {
+    pub fn verify_tx(tx: Tx) -> Result<VerifiedTx, VMError> {
+        let bp_gens = BulletproofGens::new(256, 1);
```
Comment (oleganza) [inline zkvm/src/verifier.rs]: We cannot hardcode this for two reasons:

1. Bigger txs will have larger circuits and more than 128 multipliers, failing the verification if 128 is hardcoded.
2. Computing generators is expensive. Validator would want to precompute them and use over its whole lifetime to validate arbitrary amount of transactions, so we need to pass it in.

Comment (vickiniu) [reply zkvm/src/verifier.rs]: Got it, so on the prover we'd want to calculate the capacity that we need?

If the verifier might want to precompute generators, would we need to keep `bp_gens` as a parameter on `verify_tx`?

Comment (oleganza) [reply zkvm/src/verifier.rs]: Yeah, that's why this discussion was prompted: https://dalek-cryptography.slack.com/archives/CBKMRC8DN/p1552425846018800

Comment (oleganza) [reply zkvm/src/verifier.rs]: As of exposing BP API, we could re-export only `BulletproofGens` in the lib.rs to allow user precompute them and pass into verifier. This way if they don't care about other parts of R1CS, they can avoid explicitly linking with `bulletproofs` crate.

Comment (vickiniu) [reply zkvm/src/verifier.rs]: got it, that makes sense! reading slack thread now :)

**Comment by oleganza** [comment]
We should probably simply do `pub use bulletproofs;` in the lib.rs to allow users pick various bulletproofs types (such as generators) directly via `zkvm::bulletproofs::...` to avoid adding BP as a dependency.

**Pull request closed without merging**